dask_expr/io/parquet.py

```diff
@@ -85,6 +87,128 @@ def _tokenize_fileinfo(fileinfo):
 _CACHED_PLAN_SIZE = 10
 _cached_plan = {}
 
+import pickle
+import weakref
+
+
+class FragmentWrapper:
+    _filesystems = weakref.WeakValueDictionary()
+
+    def __init__(self, fragment=None, file_size=None, fragment_packed=None) -> None:
+        """Wrap a pyarrow Fragment to only deserialize when needed."""
+        # https://github.com/apache/arrow/issues/40279
+        self._fragment = fragment
+        self._fragment_packed = fragment_packed
+        self._file_size = file_size
+        self._fs = None
+
+    def pack(self):
+        if self._fragment_packed is None:
+            self._fragment_packed = (
+                self._fragment.format,
+                (
+                    self._fragment.path
+                    if self._fragment.buffer is None
+                    else self._fragment.buffer
+                ),
+                pickle.dumps(self._fragment.filesystem),
+                self._fragment.partition_expression,
+                self._file_size,
+            )
+        self._fs = self._fragment = None
+
+    def unpack(self):
+        if self._fragment is None:
+            (
+                pqformat,
+                path_or_buffer,
+                fs_raw,
+                partition_expression,
+                file_size,
+            ) = self._fragment_packed
+            fs = FragmentWrapper._filesystems.get(fs_raw)
+            if fs is None:
+                fs = pickle.loads(fs_raw)
+                FragmentWrapper._filesystems[fs_raw] = fs
```

Comment:
Note: even if I do not deduplicate but just `loads` the filesystem, the memory leak is gone. No idea what's going on but by caching this we avoid the initialization cost, even if it's just a ms